odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,532 @@
+        i-th level detail coefficients.
+    """
+    shape = tuple(shape)
+    if any(int(s) != s for s in shape):
+        raise ValueError('`shape` may only contain integers, got {}'
+                         ''.format(shape))
+    wavelet = pywt_wavelet(wavelet)
+
+    nlevels, nlevels_in = int(nlevels), nlevels
+    if float(nlevels_in) != nlevels:
+        raise ValueError('`nlevels` must be integer, got {}'
+                         ''.format(nlevels_in))
+    # TODO: adapt for axes
+    _check_nlevels(nlevels, min(shape), wavelet)
+
+    mode, mode_in = str(mode).lower(), mode
+    if mode not in PYWT_SUPPORTED_MODES:
+        raise ValueError("mode '{}' not understood".format(mode_in))
+
+    # Use pywt.dwt_coeff_len to determine the coefficient lengths at each
+    # scale recursively. Start with the image shape and use the last created
+    # shape for the next step.
+    shape_list = [shape]
+    for i in range(nlevels):
+        shape = tuple(pywt.dwt_coeff_len(n, filter_len=wavelet.dec_len,
+                                         mode=mode)
+                      for n in shape_list[-1])
+        shape_list.append(shape)
+
+    # Add a duplicate of the last entry for the approximation coefficients
+    shape_list.append(shape_list[-1])
+
+    # We created the list in reversed order, reverse it. Remove also the
+    # superfluous image shape at the end.
+    shape_list.reverse()
+    shape_list.pop()
+    return shape_list
+
+
+def pywt_flat_coeff_size(shape, wavelet, nlevels, mode):
+    """Return the size of a flat array containing all coefficients.
+
+    Parameters
+    ----------
+    shape : sequence
+        Shape of an input to the transform.
+    wavelet : string or ``pywt.Wavelet``
+        Specification of the wavelet to be used in the transform.
+        If a string is given, it is converted to a ``pywt.Wavelet``.
+        Use `pywt.wavelist
+        <https://pywavelets.readthedocs.io/en/latest/ref/wavelets.html#\
+built-in-wavelets-wavelist>`_ to get a list of available wavelets.
+    nlevels : positive int
+        Number of scaling levels to be used in the decomposition. The
+        maximum number of levels can be calculated with
+        `pywt.dwt_max_level
+        <https://pywavelets.readthedocs.io/en/latest/ref/dwt-discrete-\
+wavelet-transform.html#maximum-decomposition-level-dwt-max-level>`_.
+    mode : string, optional
+        PyWavelets style signal extension mode. See `signal extension modes
+        <https://pywavelets.readthedocs.io/en/latest/ref/signal-extension-\
+modes.html>`_ for available options.
+
+    Returns
+    -------
+    flat_size : int
+        Size of a flat array containing all approximation and detail
+        coefficients for the specified transform.
+    """
+    shapes = pywt_coeff_shapes(shape, wavelet, nlevels, mode)
+    # 1 x approx coeff and (2**n - 1) * detail coeff
+    ndim = len(shapes[0])
+    return (np.prod(shapes[0]) +
+            sum((2 ** ndim - 1) * np.prod(shape)
```

Comment:
Also readability, I was looking for how it depended on the inner variables and couldn't find it. If it is inside you kinda expect it to be used.